## Array vs Lists

In [1]:
import numpy as np

In [2]:
l=[1,2,3]

In [3]:
A=np.array(l)
A

array([1, 2, 3])

In [4]:
for e in l:
    print(e)

1
2
3


In [5]:
for e in A:
    print(e)

1
2
3


In [7]:
l.append(4)

In [8]:
l

[1, 2, 3, 4]

In [9]:
l+[5]

[1, 2, 3, 4, 5]

## Broadcasting

In [11]:
A+np.array([4])

array([5, 6, 7])

In [12]:
A+np.array([5,7,8])

array([ 6,  9, 11])

In [13]:
A+np.array([4,3]) # sizes are different

ValueError: operands could not be broadcast together with shapes (3,) (2,) 

In [14]:
2*A

array([2, 4, 6])

In [15]:
2*l

[1, 2, 3, 4, 1, 2, 3, 4]

In [16]:
l+l

[1, 2, 3, 4, 1, 2, 3, 4]

In [17]:
l2=[]
for e in l:
    l2.append(e+3)
    
l2

[4, 5, 6, 7]

In [19]:
l2=[e+3 for e in l]
l2

[4, 5, 6, 7]

In [20]:
l**2 # power doesnt support l2

TypeError: unsupported operand type(s) for ** or pow(): 'list' and 'int'

In [21]:
l2=[]
for e in l:
    l2.append(e**2)

l2

[1, 4, 9, 16]

In [22]:
A**2# if we have a array it works perfectly fine

array([1, 4, 9], dtype=int32)

In [23]:
np.log(A)

array([0.        , 0.69314718, 1.09861229])

In [24]:
np.sum(A)

6

In [25]:
np.sqrt(A)

array([1.        , 1.41421356, 1.73205081])

In [26]:
np.exp(A)

array([ 2.71828183,  7.3890561 , 20.08553692])

In [27]:
np.tanh(A)

array([0.76159416, 0.96402758, 0.99505475])

## DOT product

In [28]:
a=np.array([1,2])
b=np.array([3,4])

In [33]:
dot=0
for e, f in zip(a,b):
    dot+=e+f

In [34]:
dot

10

In [36]:
dot=0
for i in range(len(a)):
    dot+=a[i]*b[i]
dot

11

In [38]:
a*b#element wise multiplication

array([3, 8])

In [39]:
np.sum(a*b)

11

In [40]:
a*b.sum()

array([ 7, 14])

In [41]:
np.dot(a,b)

11

In [42]:
a.dot(b)

11

In [44]:
a @ b# the symbol is for the dot product

11

## Dot Product

In [45]:
amag=np.sqrt((a*a).sum())

In [46]:
amag

2.23606797749979

In [48]:
np.linalg.norm(a)

2.23606797749979

In [49]:
cosangle=a.dot(b)/(np.linalg.norm(a)*np.linalg.norm(b))

In [50]:
cosangle

0.9838699100999074

In [53]:
angle=np.arccos(cosangle)
angle # in radians

0.17985349979247847

## Metrices:

In [54]:
L=[[1,2],[3,4]]
L

[[1, 2], [3, 4]]

In [55]:
L[0]

[1, 2]

In [56]:
L[1]

[3, 4]

In [58]:
L[0][1]

2

In [59]:
A=np.array([[1,2],[3,4]])

In [60]:
A

array([[1, 2],
       [3, 4]])

In [61]:
A[0][1]

2

In [62]:
A[0,1]#otherwise to notate

2

In [63]:
A[:,0]

array([1, 3])

In [67]:
A[:,1]

array([2, 4])

In [68]:
np.exp(A)

array([[ 2.71828183,  7.3890561 ],
       [20.08553692, 54.59815003]])

In [69]:
np.exp(L)

array([[ 2.71828183,  7.3890561 ],
       [20.08553692, 54.59815003]])

## Matrix operation 

In [70]:
B=np.array([[1,2,3],[4,5,6]])
B

array([[1, 2, 3],
       [4, 5, 6]])

In [71]:
A.dot(B)

array([[ 9, 12, 15],
       [19, 26, 33]])

In [72]:
A.dot(B.T)

ValueError: shapes (2,2) and (3,2) not aligned: 2 (dim 1) != 3 (dim 0)

In [73]:
np.linalg.det(A)

-2.0000000000000004

In [74]:
np.linalg.inv(A)

array([[-2. ,  1. ],
       [ 1.5, -0.5]])

In [76]:
#A* A inverse should give identity

A.dot(np.linalg.inv(A))

array([[1.00000000e+00, 1.11022302e-16],
       [0.00000000e+00, 1.00000000e+00]])

In [78]:
np.linalg.inv(A).dot(A)# A is inaccurate a bit so you will not getting the exact zero, but close to zero

array([[1.0000000e+00, 4.4408921e-16],
       [0.0000000e+00, 1.0000000e+00]])

In [79]:
np.trace(A)

5

In [80]:
np.diag(A)

array([1, 4])

In [81]:
np.diag([1,4])

array([[1, 0],
       [0, 4]])

## Eigen values & Eigen Vectors 

In [82]:
np.linalg.eig(A)

(array([-0.37228132,  5.37228132]), array([[-0.82456484, -0.41597356],
        [ 0.56576746, -0.90937671]]))

In [86]:
lam, V=np.linalg.eig(A)

In [88]:
V[:,0]*lam[0]==A@V[:,0]

array([ True, False])

In [89]:
V[:,0]*lam[0], A@V[:,0]

(array([ 0.30697009, -0.21062466]), array([ 0.30697009, -0.21062466]))

In [92]:
np.allclose(V[:,0]*lam[0],A @ V[:,0])

True

In [96]:
V.shape,lam.shape,A.shape

((2, 2), (2,), (2, 2))

In [97]:
V,lam,A

(array([[-0.82456484, -0.41597356],
        [ 0.56576746, -0.90937671]]),
 array([-0.37228132,  5.37228132]),
 array([[1, 2],
        [3, 4]]))

In [101]:
np.allclose(V@np.diag(lam),A@V)

True

## Liner Systems

The admission fee at a small fair is  '$1.50'     for children  and      '$4.00' for adults. on a certain day, 2200 people enter the fair, and    $5500 is collected. How many children and how many adults attended?

2 equations * 2 unknows

This can be posed as 


x1+x2 = 2200

1.5x1+4X2=5500

x=(x1,x2) A=((1,1)(1.5,4)) , b=(2200,5500)


Ax=b, x=A-1x

## don't do that literally 

1. The inverse is slower and less accurate 
2. There are better algorithms to slove liner systems. 

x1=np.linalg.solve(A,b) #yes


x2=np.linalg.inv(A).dot(b) #no 

linalg.solve # will be more accurate and fast

In [108]:
A=np.array([[1,1],[1.5,4]])
b=np.array([2200,5050])

In [111]:
np.linalg.solve(A,b)# you need to go for solve function

array([1500.,  700.])

In [110]:
np.linalg.inv(A).dot(b)# this is slower

array([1500.,  700.])

## Generate Data

In [112]:
np.zeros((2,3))

array([[0., 0., 0.],
       [0., 0., 0.]])

In [113]:
np.ones((2,3))

array([[1., 1., 1.],
       [1., 1., 1.]])

In [115]:
10*np.ones((2,2))

array([[10., 10.],
       [10., 10.]])

In [116]:
np.eye(3)

array([[1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.]])

In [124]:
np.random.random((2,3))

array([[0.73216602, 0.67395312, 0.21379234],
       [0.70299976, 0.12779555, 0.14790792]])

In [131]:
np.random.rand(10,2)

array([[0.06578808, 0.25828329],
       [0.03963143, 0.91539489],
       [0.8671334 , 0.06847458],
       [0.4300427 , 0.3129784 ],
       [0.30840151, 0.36950463],
       [0.38622839, 0.82654725],
       [0.64099893, 0.66457948],
       [0.93440019, 0.79058687],
       [0.92579264, 0.79070263],
       [0.64894758, 0.43199373]])

In [144]:
#randn gives Return a sample (or samples) from the "standard normal" distribution.
r=np.random.randn(10)
r.mean()

-0.3769339604344242

In [145]:
np.random.randn(10).mean()

0.0011956925292743823

In [146]:
R=np.random.randn(1000,3)

In [147]:
R.mean(axis=0)

array([-0.03449739,  0.05222461, -0.02496412])

In [148]:
R.mean(axis=1)

array([ 4.93280578e-01, -3.52502723e-01,  8.20319498e-02, -5.42172148e-01,
       -4.67099713e-01, -4.25133826e-01,  3.23639554e-01, -1.80691602e-02,
       -9.81686519e-02,  7.43503687e-01,  2.02068297e-01, -4.68681149e-01,
       -2.96750927e-01,  1.12518094e-02, -4.79019688e-01,  1.75905126e+00,
       -3.75933222e-01,  2.40845572e-01,  3.17013089e-01,  4.79284483e-01,
        5.17347427e-02, -3.49820883e-01,  4.61313060e-01, -1.75684008e-01,
       -5.00236829e-01,  6.20517710e-02,  6.23890851e-02, -8.69804999e-01,
        7.38774809e-01,  4.79579339e-01, -7.89458358e-01, -2.52412499e-01,
        7.62201144e-02,  8.58617623e-01, -8.86150337e-02,  3.45346005e-01,
       -8.11264331e-03,  6.36963630e-01, -3.79356897e-01, -1.33900902e-01,
        5.24321505e-01, -1.61620955e+00,  4.74606804e-01,  1.15131983e+00,
       -1.03977901e+00,  3.06580977e-01, -3.85076478e-01,  3.23445093e-01,
        4.10391096e-01, -5.26381680e-01,  3.47490896e-02, -6.01594989e-01,
       -2.98527191e-01, -

In [149]:
R.mean(axis=1).shape

(1000,)

In [150]:
np.cov(R).shape

(1000, 1000)

In [151]:
np.cov(R.T)

array([[ 0.96064771, -0.00267718, -0.07176531],
       [-0.00267718,  1.00208678, -0.01182504],
       [-0.07176531, -0.01182504,  0.99385182]])

In [152]:
np.cov(R, rowvar=False)

array([[ 0.96064771, -0.00267718, -0.07176531],
       [-0.00267718,  1.00208678, -0.01182504],
       [-0.07176531, -0.01182504,  0.99385182]])

In [154]:
np.random.randint(0,20,size=(10,3))

array([[18, 15,  2],
       [ 8, 11, 14],
       [15,  3, 12],
       [ 8,  1,  4],
       [18,  4,  3],
       [12, 15, 16],
       [ 1,  3, 19],
       [ 6, 14,  1],
       [ 0, 19, 18],
       [ 9,  6, 16]])

In [157]:
np.random.choice(15,size=(3,3),replace=False)

array([[ 4,  3,  9],
       [ 8, 10,  6],
       [ 2,  5, 13]])